Tyler test branch

tylerdvanl opened the pull request:
I'm wondering how we want to approach logging and payment.  Professor doesn't like when objects that aren't the main method spit out to console, and if we want pay() to be called inside of RoachMotel, we'll need to add a member variable to a class to store what payment method to use.

tylerdvanl committed:
PaymentMethods

added constructors and member variables for Payment Methods

tylerdvanl committed:
RoachMotel.pay()

A question about how we want to handle logging and payment.

thaolydo commented on src/cecs277TermProject/RoachMotel.java:
```diff
@@ -52,6 +51,19 @@ public void checkOut(RoachColony roachColony) {
+	 * @param paymentMethod  A payment method (RoachPal or MasterRoach) passed in through the main method.
+	 * @param cost The total cost to be paid.
+	 */
+	public void pay(PaymentStrategy paymentMethod, double cost)
```
This should be private and used only in `checkout()` where the params will be: colony, numberOfDays, and payment method.

tylerdvanl replied:
Do you want me to calculate the cost in that method?

thaolydo replied:
yes.

tylerdvanl replied:
do we have a way to tell what room a colony is staying in yet?

thaolydo commented:
> I'm wondering how we want to approach logging and payment. Professor doesn't like when objects that aren't the main method spit out to console

Yeah, let's move the `System.out.println()` from `checkIn()` and `checkOut()` to `main()`. As for logging, we shoud log to a file. We can do that by creating a new `Logger` class that has a `log()` method and using Singleton pattern to create it in `RoachMotel.java` and save it as a member variable. We will use the logger instance (member variable) to log events in `checkIn()`, `checkOut()`, and `pay()`. In `main()`, we will use logger instance to get the file name, and use that to read the log files and print it out.

> and if we want pay() to be called inside of RoachMotel, we'll need to add a member variable to a class to store what payment method to use.

The payment method object should be created in `main()` and pass to `checkOut()` as parameter.

tylerdvanl committed:
RoachMotel.pay()

Changed arguments, needs additional info from Room and Motel in order to work properly

tylerdvanl committed:
Better version of pay()

This should work no matter how getCost() works now; next up is logging i believe

tylerdvanl committed:
Basic Logger

According to the project spec, the log has a different format than normal console, so we need to have methods that will write strings in that format for use in the logger.

thaolydo commented on src/cecs277TermProject/Logger.java:
```diff
@@ -0,0 +1,28 @@
+	private File outFile;
+	private PrintWriter out;
+	
+	public Logger(File outFile) throws FileNotFoundException
```
outFile should be initialized in here instead of let the caller side instantiate it. This constructor should take no argument.

thaolydo commented on src/cecs277TermProject/Logger.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Logger is simply our file I/O class, meant to write to the transaction log.
+ */
+public class Logger 
```
This class should be a singleton. You should implement singleton pattern for this class.

thaolydo commented on src/cecs277TermProject/Logger.java:
```diff
@@ -0,0 +1,28 @@
+{
+	
+	private File outFile;
+	private PrintWriter out;
```
There should be a fileName variable that store log file name, and expose a getter for that variable so that in main() we can read the log file from the file name in this class.

thaolydo commented on src/cecs277TermProject/RoachMotel.java:
```diff
@@ -16,13 +18,15 @@
 
 	// TODO: [Gust] Make this constructor private and use it in Singleton pattern
-	public RoachMotel(String name, int capacity) {
+	public RoachMotel(String name, int capacity, File outFile) throws FileNotFoundException {
```
As comment above, outFile should be instantiate in Logger.java

thaolydo commented on src/cecs277TermProject/MotelRunner.java:
```diff
@@ -1,10 +1,20 @@
+		RoachMotel motel = null;
+		try
+		{
+			motel = new RoachMotel("The only motel in town", 6, null); // TODO: Use Singleton pattern here
```
As commented above, outFile should be instantiated in Logger.java

thaolydo commented on src/cecs277TermProject/RoachMotel.java:
```diff
@@ -45,14 +49,13 @@ public void checkIn(RoachColony roachColony) {
-	public void checkOut(RoachColony roachColony) {
+	public void checkOut(RoachColony roachColony, PaymentStrategy paymentMethod, int numDays) {
 		// TODO: [Tyler] Implement logging this event here
 		System.out.printf("Checking out: %s\n", roachColony);
```
Please add a log statement here.

thaolydo commented on src/cecs277TermProject/RoachMotel.java:
```diff
@@ -61,9 +64,12 @@ public void checkOut(RoachColony roachColony) {
+	private String pay(PaymentStrategy paymentMethod, RoachColony colony, int numDays)
 	{
-		paymentMethod.pay(cost);
+		Room paidRoom = occupiedRooms.get(colony);
```
Add a log statement here.

tylerdvanl committed:
payment ToString; Logging

PaymentStrategies have toStrings now

Logging should work; there is a small amount of placeholder text, waiting for room code to be finished before I touch those up

tylerdvanl committed:
Singleton, outfile Instantiation

Logger should now be using singleton design pattern, and outfile is now created inside the constructor

tylerdvanl committed:
filename, getFilename

filename is now a variable, and has a getter

thaolydo commented on src/cecs277TermProject/Logger.java:
```diff
@@ -9,13 +9,31 @@
+	{
+		this.filename = "roach_motel_log.txt";
+		this.outFile = new File(filename);
 		out = new PrintWriter(outFile);
```
I think we need to expose a method to close the writer (I could be wrong) and use it in main after we done logging (after checking in/out and before printing out the log in `main()`).

tylerdvanl replied:
I changed it open and close a printwriter during the log() function

thaolydo commented on src/cecs277TermProject/Logger.java:
```diff
@@ -24,5 +42,9 @@ public void log(String message)
 		out.println(message);
 	}
 	
+	public String getFilename()
+	{
+		return this.filename;
+	}
```
Now that I think about it, we should have a method that return the `File` instance itself (called it `inFile` as member variable), instead of just file name. It will make it easier for `main()` to start reading the log file.

thaolydo commented on src/cecs277TermProject/RoachMotel.java:
```diff
@@ -22,11 +22,11 @@
 		this.capacity = capacity;
 		this.occupiedRooms = new HashMap<>();
-		this.logger = new Logger(outFile);
+		this.logger = logger.getLogger();
```
Nice!

thaolydo commented on src/cecs277TermProject/RoachMotel.java:
```diff
@@ -41,20 +41,21 @@ public RoachMotel(String name, int capacity, File outFile) throws FileNotFoundEx
 
 		// TODO: [Gust] Implement decorator pattern here for amenities before associating the colony with the room
 		occupiedRooms.put(roachColony, room);
+		logCheckIn(roachColony, this.logger);
```
logging should be at the beginning of a method (to keep track of of what events happens). If it's at the end, there maybe exception, and the event would be never logged. You can put a log here as well to log the successful registration event. I don't think we need to create a separate method `logCheckIn()` as log statements should be 1-line statements.

tylerdvanl replied:
I think we need separate methods for checkIn and checkOut because when they check out the system needs to log how they paid, but when checking in they dont.

thaolydo commented on src/cecs277TermProject/RoachMotel.java:
```diff
@@ -41,20 +41,21 @@ public RoachMotel(String name, int capacity, File outFile) throws FileNotFoundEx
 	// TODO: [Gust] Add a parameter for list of amenities
 	public void checkIn(RoachColony roachColony) {
 		// TODO: [Tyler] Implement logging this event here
+		// Logging should be mostly finished; just need room methods to be finished - Tyler
```
Which room methods are you waiting for? I think you only need the `getCost()` method (which already merged in master). Gust will take care of implementing Decorator pattern, which include adding amenities, but the interface should not change. Am I missing something?

tylerdvanl replied:
I only saw room number in the Room.toString().  If that's gonna print out amenities too I can get rid of my placeholder text and it should be working completely then.

thaolydo commented on src/cecs277TermProject/RoachMotel.java:
```diff
@@ -69,8 +70,26 @@ private String pay(PaymentStrategy paymentMethod, RoachColony colony, int numDay
+	{
+		Room loggedRoom = occupiedRooms.get(colony);
+		
+		log.log("Successfully Checked In: Colony " + colony + "checking into " + loggedRoom + "with amenities [PLACEHOLDER]");
```
Use `String.format()` as it's more readable. Also, `loggedRoom` should include amenities (to-be-implemented by Gust), so you don't need to log them here.

thaolydo commented on src/cecs277TermProject/RoachMotel.java:
```diff
@@ -69,8 +70,26 @@ private String pay(PaymentStrategy paymentMethod, RoachColony colony, int numDay
+	{
+		Room loggedRoom = occupiedRooms.get(colony);
+		
+		log.log("Successfully Checked Out: Colony" + colony + " checking out of " + loggedRoom + " with amenities [PLACEHOLDER] "
```
Same as above.

thaolydo commented on src/cecs277TermProject/RoachMotel.java:
```diff
@@ -41,20 +41,21 @@ public RoachMotel(String name, int capacity, File outFile) throws FileNotFoundEx
 
 	public void checkOut(RoachColony roachColony, PaymentStrategy paymentMethod, int numDays) {
 		// TODO: [Tyler] Implement logging this event here
+		// Logging should be mostly finished; just need room methods to be finished - Tyler
```
same as above

thaolydo commented on src/cecs277TermProject/RoachMotel.java:
```diff
@@ -69,8 +70,26 @@ private String pay(PaymentStrategy paymentMethod, RoachColony colony, int numDay
 		Room paidRoom = occupiedRooms.get(colony);
 		// in progress, needs a way to calculate cost from amenities - tyler
 		double cost = paidRoom.getCost() * numDays;
+		logCheckOut(paymentMethod, colony, numDays, this.logger);
```
same as above in `checkIn()`.

thaolydo commented:
Can you please also do a rebase (or merge from master if you're comfortable with rebase) so that it will be easier when we merge this pull request and also the diff is easier to review as well. There will be some conflicts, but you can resolve them.

tylerdvanl committed:
Various Fixes

-log() now creates and closes it's own printwriter

-log happens at beginning of checkIn and checkOut

-changed things to use String.format

tylerdvanl commented:
Did a rebase, I think i did it properly, please look it over?

thaolydo commented on src/cecs277TermProject/Logger.java:
```diff
@@ -0,0 +1,65 @@
+	{
+		try 
+		{
+			PrintWriter out = new PrintWriter(outFile);
```
Move this variable to be a member var.

thaolydo commented on src/cecs277TermProject/Logger.java:
```diff
@@ -0,0 +1,65 @@
+		try 
+		{
+			PrintWriter out = new PrintWriter(outFile);
+			out.println(message);
```
It'll be great if you can add timestamp as a prefix to the message if it's not too much work.

tylerdvanl replied:
yeah i forgot about that, will do

thaolydo commented on src/cecs277TermProject/MotelRunner.java:
```diff
@@ -1,58 +1,68 @@
+			System.out.println(motel);
+			System.out.println();
+
+			//motel.checkOut(colony2);
```
why is this line commented out? We want to test `checkOut()` method as well.

tylerdvanl replied:
mistake, was trying things out, forgot to change it back

thaolydo reviewed (changes requested):
For logging, the practice is 1 log statement per message you want to log; you don't need to create new methods (`logCheckIn()` and `logCheckOut()`) for logging in `RoachMotel.java`. If the message to log in `checkIn()` is different from message in `checkOut()`, you just need to pass in a different message string (using `String.format` in logging will help).

tylerdvanl committed:
moar fixes

-removed logcheckin and logcheckout, replaced with calls to log.

-logger.out moved to member variable, still opened and closed within log()

-checkIn logs all attempts, successes, and failures.

-Added timestamps to logs(I think this works)

thaolydo commented on src/cecs277TermProject/Logger.java:
```diff
@@ -34,15 +37,19 @@ private Logger() throws FileNotFoundException
 		{
-			PrintWriter out = new PrintWriter(outFile);
-			out.println(message);
+			Instant time = Instant.now();
+			DateTimeFormatter format = DateTimeFormatter.ofPattern("uuuu-MM-dd HH:mm:ss:SSS");
+			String timestamp = format.format(time);
```
nice time format!

thaolydo commented on src/cecs277TermProject/Logger.java:
```diff
@@ -34,15 +37,19 @@ private Logger() throws FileNotFoundException
+			Instant time = Instant.now();
+			DateTimeFormatter format = DateTimeFormatter.ofPattern("uuuu-MM-dd HH:mm:ss:SSS");
+			String timestamp = format.format(time);
+			out = new PrintWriter(outFile);
```
Move this to the constructor because this method will be called a lot and we don't want to create a new PrintWriter object everytime.

thaolydo commented on src/cecs277TermProject/MotelRunner.java:
```diff
@@ -23,7 +23,9 @@ public static void main(String[] args) {
-			//motel.checkOut(colony2);
+			
+			//When we decide to finish testing everything, this needs to have a payment method added as an argument. - Tyler
+			motel.checkOut(colony2);
```
Pass a new payment method object and the number of day in here as well.

tylerdvanl committed:
Updated Logger

-Added close method, to be called at the end of main

-Checkout test in main compiles

-moved printwriter again

thaolydo commented on src/cecs277TermProject/Logger.java:
```diff
@@ -0,0 +1,72 @@
+			String timestamp = format.format(time);
+			out = new PrintWriter(outFile);
+			out.println(String.format("[%1$s] %2$s", timestamp, message));
+			out.close();
```
I see that you open and close them for every log() invocation. I guess it's ok for this project, but it's an expensive operation.

thaolydo commented on src/cecs277TermProject/MotelRunner.java:
```diff
@@ -1,58 +1,68 @@
+			System.out.println(motel);
+
+			// Printing the log
+			// TODO: implement reading and printing the logs
```
We can start a new pull request for this task.
